flang/include/flang/Optimizer/Builder/BoxValue.h

```diff
@@ -410,14 +411,36 @@ bool isArray(const ExtendedValue &exv);
 /// Get the type parameters for `exv`.
 llvm::SmallVector<mlir::Value> getTypeParams(const ExtendedValue &exv);
 
+//===----------------------------------------------------------------------===//
+// Functions that may generate IR to recover properties from extended values.
+//===----------------------------------------------------------------------===//
+namespace factory {
+
+/// Generalized function to recover dependent type parameters. This does away
+/// with the distinction between deferred and non-deferred LEN type parameters
+/// (Fortran definition), since that categorization is irrelevant when getting
+/// all type parameters for a value of dependent type.
+llvm::SmallVector<mlir::Value> getTypeParams(mlir::Location loc,
+                                             FirOpBuilder &builder,
+                                             const ExtendedValue &exv);
+
```

Comment:
I am not against moving these helpers from `FirBuilder.h` to `BoxValue.h` and swapping the factory helper arguments order from `(builder, loc, args)` to `(loc, builder, args)` (I used the former because it remained me of the order in `builder->create<>(loc, args)`, but I am not attached to it).

Having these helpers here makes actually sense to me, and having `loc` first is in line with most other functions, so it the change sounds good to me.

However, if this is to be done, I think this should be done for all the factory functions doing things with BoxValue, otherwise we will not know where to look for a helpers, and arguments order consistency seems more important to me than the actual order. So I would suggest doing a NFC patch touching all such functions.

Reply:
MLIR went through an exercise a while back to standardize Location arguments. We should aspire to use the same convention.

Reply:
I agree that we should consolidate factory functions that generate code relative to a bridge box value here.

This is a start and I'll do more refactoring in a future PR.

Why? It is easier to find the right calls if highly-related code is all in the same place. Having some of the code that deals with bridge boxes here while other bits with somewhat different names, etc. in other places makes the code harder to maintain.